homeassistant/components/spotify/media_player.py

```diff
@@ -60,332 +58,264 @@
-        return self._image_url
+        if (
+            self._currently_playing.get("item") is None
+            or not self._currently_playing["item"]["album"]["images"]
+        ):
+            return None
+        return self._currently_playing["item"]["album"]["images"][0]["url"]
+
+    @property
+    def media_image_remotely_accessible(self) -> bool:
+        """If the image url is remotely accessible."""
+        return False
+
+    @property
+    def media_title(self) -> Optional[str]:
+        """Return the media title."""
+        return self._currently_playing.get("item", {}).get("name")
 
     @property
-    def media_artist(self):
+    def media_artist(self) -> Optional[str]:
         """Return the media artist."""
-        return self._artist
+        if self._currently_playing.get("item") is None:
+            return None
+        return ", ".join(
+            [artist["name"] for artist in self._currently_playing["item"]["artists"]]
+        )
 
     @property
-    def media_album_name(self):
+    def media_album_name(self) -> Optional[str]:
         """Return the media album."""
-        return self._album
+        if self._currently_playing.get("item") is None:
+            return None
+        return self._currently_playing["item"]["album"]["name"]
 
     @property
-    def media_title(self):
-        """Return the media title."""
-        return self._title
+    def media_track(self) -> Optional[int]:
+        """Track number of current playing media, music track only."""
+        return self._currently_playing.get("item", {}).get("track_number")
 
     @property
-    def supported_features(self):
-        """Return the media player features that are supported."""
-        if self._user is not None and self._user["product"] == "premium":
-            return SUPPORT_SPOTIFY
-        return None
+    def media_playlist(self):
+        """Title of Playlist currently playing."""
+        if self._playlist is None:
+            return None
+        return self._playlist["name"]
 
     @property
-    def media_content_type(self):
-        """Return the media type."""
-        return MEDIA_TYPE_MUSIC
+    def source(self) -> Optional[str]:
+        """Return the current playback device."""
+        return self._currently_playing.get("device", {}).get("name")
+
+    @property
+    def source_list(self) -> Optional[List[str]]:
+        """Return a list of source devices."""
+        if not self._devices:
+            return None
+        return [device["name"] for device in self._devices]
+
+    @property
+    def shuffle(self) -> bool:
+        """Shuffling state."""
+        return bool(self._currently_playing.get("shuffle_state"))
+
+    @property
+    def supported_features(self) -> int:
+        """Return the media player features that are supported."""
+        if (
+            self._me is not None and self._me["product"] != "premium"
+        ) or self._currently_playing.get("device", {}).get("is_restricted", True):
```

Comment:
Can the device change during entity lifetime? Normally we don't want the supported features to change during entity lifetime.

Reply:
yes, the selected source defines the capabilities.

Reply:
Well, the supported features should not change during entity lifetime according to our decision here:
https://github.com/home-assistant/architecture/issues/1#issuecomment-370202970

Reply:
OK, thanks, in that case, there is no way to implement this correctly and will suggest removing the Spotify integration completely.

Reply:
Maybe there could be a callback when a new device connects/disconnects that creates/removes the entity?

Reply:
That sounds really not wished for. Spotify is very dynamic in its available sources. e.g., open a browser player: new device. Turn on your Spotify enabled TV new device. Close browser tab, device gone.

Reply:
Implementation sample/suggestion @ media player entity component:

```py
    @property
    def supported_features(self):
        """Flag media player features that are supported."""
        return 0

    @property
    def active_supported_features(self):
        """Flag media player features that are supported and active."""
        return self.supported_features
```

Reply:
This discussion should probably move to the architecture repo.

Reply:
You are absolutely right, I'll do that later today. 👍

Reply:
Proposal is up: <https://github.com/home-assistant/architecture/issues/338>

Reply:
Based on the results of the proposal and my own tinkering this weekend; I'm going to change the implementation.

Reply:
Resolved. The Spotify client is initialized first, before the platform.
The supported features are now solely based on the account type, which is known before the creation of the entity and thus static during runtime.